odl/solvers/functional/default_functionals.py

```diff
@@ -2292,6 +2294,219 @@ def gradient(self):
+            ``gamma > 0``, it has a ``1/gamma``-Lipschitz gradient so that
+            its convex conjugate is ``gamma``-strongly convex.
+
+        Examples
+        --------
+        Example of initializing the Huber norm functional
+
+        >>> space = odl.uniform_discr(0, 1, 14)
+        >>> gamma = 0.1
+        >>> huber_norm = odl.solvers.Huber(space, gamma=0.1)
+
+        Check that if all elements are > ``gamma`` we get the L1-norm up to a
+        constant.
+
+        >>> x = 2 * gamma * space.one()
+        >>> tol = 1e-5
+        >>> constant = gamma / 2 * space.one().inner(space.one())
+        >>> l1_norm = odl.solvers.L1Norm(space) - constant
+        >>> abs(huber_norm(x) - l1_norm(x)) < tol
+        True
+
+        Check that if all elements are < ``gamma`` we get the squared L2-norm
+        times the weight 1/(2*gamma).
+
+        >>> x = gamma / 2 * space.one()
+        >>> l2_norm = 1 / (2 * gamma) * odl.solvers.L2NormSquared(space)
+        >>> abs(huber_norm(x) - l2_norm(x)) < tol
+        True
+
+        Compare Huber- and L1-norm for vanishing smoothing ``gamma=0``.
+
+        >>> x = odl.phantom.white_noise(space)
+        >>> huber_norm = odl.solvers.Huber(space, gamma=0)
+        >>> l1_norm = odl.solvers.L1Norm(space)
+        >>> abs(huber_norm(x) - l1_norm(x)) < tol
+        True
+
+        Redo previous example for a product space in two dimensions.
+
+        >>> domain = odl.uniform_discr([0, 0], [1, 1], [5, 5])
+        >>> space = odl.ProductSpace(domain, 2)
+        >>> x = odl.phantom.white_noise(space)
+        >>> huber_norm = odl.solvers.Huber(space, gamma=0)
+        >>> l1_norm = odl.solvers.GroupL1Norm(space, 2)
+        >>> abs(huber_norm(x) - l1_norm(x)) < tol
+        True
+        """
+
+        self.__gamma = float(gamma)
+        self.strong_convexity = 0
+
+        if self.gamma > 0:
+            grad_lipschitz = 1 / self.gamma
+        else:
+            grad_lipschitz = np.inf
+
+        super(Huber, self).__init__(space=space, linear=False,
+                                    grad_lipschitz=grad_lipschitz)
+
+    @property
+    def gamma(self):
+        """The smoothing parameter of the Huber norm functional."""
+        return self.__gamma
+
+    def local_norm(self, x):
+        if isinstance(self.domain, ProductSpace):
+            return PointwiseNorm(self.domain, 2)(x)
+        else:
+            return x.ufuncs.absolute()
+
+    def local_norm_operator(self):
+        if isinstance(self.domain, ProductSpace):
+            return GroupL1Norm(self.domain, 2)
+        else:
+            return L1Norm(self.domain)
+
+    def _call(self, x):
+        """Return ``self(x)``."""
+        n = self.local_norm(x)
+
+        if self.gamma > 0:
+            i = n.ufuncs.less(self.gamma)
+            out = i * 1 / (2 * self.gamma) * n**2
+
+            i.ufuncs.logical_not(out=i)
+            out += i * (n - self.gamma / 2)
+        else:
+            out = n
```

Comment:
This is a matter of taste, but a good principle for local variable names is "the larger the scope, the more verbose the name". 
In practice that means that names like `i` and `n` are fine for short loops or list comprehensions. As soon as the scope spans more than, say, 5 lines of code, though, one-letter names make code less readable simply because they are harder to distinguish from numbers, e.g. `i` or `l` from `1`.

Here, `n` could easily be `norm` without making lines annoyingly long. Also, the `i` multiplication can be replaced by boolean array indexing by going through Numpy:
```python
    norm = self.local_norm(x).asarray()
    with writable_array(out) as out_arr:
        if self.gamma > 0:
            # Quadratic part
            i = (norm < gamma)
            out_arr[i] = norm[i] ** 2 / (2 * self.gamma)
            # Absolute value part
            np.logical_not(i, out=i)
            out_arr[i] = norm[i] - self.gamma / 2
```
This is slightly slower due to boolean indexing, but more memory-friendly. You can scrape off even more copies by changing `norm` in-place before assigning to `out`, at the expense of an additional (slow) boolean indexing each time.

Reply:
I agree with the variable names. When you write this code for yourself I guess one is just too lazy :)
 
About the computations, I am not so sure. I would like to keep the code numpy independent so that you can also use it on the gpu which I believe should be the case at the moment.

Reply:
Fair enough. Just be aware that most of these ufuncs are not implemented in the CUDA backend and fall back to "copy to CPU, then use Numpy, then copy back to GPU". So you will probably end up being slower AND use more memory.

That said, Jonas is doing the (I hope) final round of review of #1088 which will make all this work much nicer.